standards/bulk.md

```diff
@@ -0,0 +1,134 @@
+# Bulk Operations
+
+## Overview
+
+The concept of bulk operations on RESTful endpoints is quite foreign, in the sense that it is not a common practice. However, there are some use cases where it makes sense to allow a client to perform multiple operations in a single request. For example, a client may want to create multiple entities at once, or update multiple entities at once. In these cases, it is often more efficient to allow the client to perform these operations in a single request, rather than making multiple requests. However, before adding support for bulk operations you should think twice if this feature is really needed. Often network performance is not what limits request throughput. Similarly, sending multiple in-parallel requests for standard REST endpoints can be a fine solution for some clients.
+
+### Synchronous
+
+Bulk operations **MUST** be synchronous when applied to an existing resource. This means that the client will receive a response to the bulk request only after all operations have been completed. This is in contrast to asynchronous operations, where the client receives a response immediately after the request is received, and then must poll for the result of the operation. Asynchronous bulk operations are applied to new resources specifically for bulk or import to enable subsequent status updates through additional endpoints.
+
+- Bulk operations **MUST** be specific to a single resource type and NOT allow for updating multiple resource types in a single request.
+- Bulk operations **MUST** be implemented as a PATCH request against a collection resource.
+- Bulk operations **MUST** return a `200 (OK)` response code if all operations were received and a result is available for each operation, regardless of success. Bulk operations **MUST NOT** use status code `207 (Multi-Status)` response code as this incurs other implications for the response schema in relationship to Web DAV.
+- Bulk operations **MUST** accept a constrained number of operations in the request body that is indicated in the documentation of the endpoint. By default this value **MAY** be 100 operations, but should be adjusting according to the needs of the endpoint and the entity-size. Requests beyond the limit **MUST** return a `400` response code and standard [error format body](errors.md) similar to:
+```
+// RESPONSE
+HTTP/1.1 400
+Content-Type: application/problem+json
+{
+    "title": "Invalid Data",
+    "status": 400,
+    "detail": "Operations collection may only contain a maximum of '100' actions per request.",
+    "instance": "/articles",
+    "requestId": "b6d9a290-9f20-465b-bcd3-4a5166eeb3d7"
+}
+```
+
+- Bulk operations **MUST** include a request body schema following: 
+```json
+{
+    "operations": [
+        {
+            "action": "CREATE" | "UPDATE" | "CREATE_UPDATE" | "DELETE",     // indicate intent of operation
```

Comment:
thoughts on these actions and terms used? do we need a "CREATE_UPDATE" regularly"

Reply:
Sounds like we may need a `FETCH` (or something like that) as well.

Reply:
Since this is resource specific why would you use the bulk PATCH endpoint to fetch instead of the normal GET request? Unless I am misunderstanding what you mean by FETCH?

Reply:
This may be outside the scope of this PR but, with Items API `/items/views/lookup` endpoint we are using an "intelligent" or "contextual" search.  This was sort of modeled after Rules Managements `/types/{id}/rules/instances/evaluate` endpoint.  Both of with are POSTs and they perform a query on a set of resources that can't be modeled (or at least not efficiently) by FIQL.  This is a pattern that we might want to elsewhere like the new Objects API.

So as it relates to this PR, we may want to do a bulk "lookup" of Items.  This falls somewhere between an action and a GET.  I don't remember the API design review discussion around the `/lookup` endpoint so I might be missing some context.

https://app.swaggerhub.com/apis/spscommerce/items-api/1.0.510#/Item%20Views/item-views-lookup

Reply:
In the Rules Management case, I think it was made a POST because it wasn't a strict drill down filter like you'd do in a normal GET, and there were some validation requirements on the inputs. GET endpoints can only have optional query parameters, so we used a POST request body.

But yeah, to your point, it's also kind of an action that happens to return a collection of results?

Reply:
CREATE_UPDATE (or "upsert") is a fairly common action, I'd say.

Reply:
Core question to answer - can we just use GET for bulk retrieval... should this support that too?
No concerns in recognizing that the references you have Nick/Evan are likely outside of this pattern and intentionally so.

Reply:
I would expect bulk retrieval to be covered by FIQL `=in=` on the collection endpoint of some resource.

Reply:
FIQL doesn't really fit the use case we're describing, however it should be sufficient in a general bulk retrieval.  As discussed in the review, bulk operations of actions will not be supported as they represent a different use case.

Reply:
It is still not clear to me if we are say that an API _must_ support _every_ action "CREATE" | "UPDATE" | "CREATE_UPDATE" | "DELETE" or that it _can_ support _any_ of these actions.

Reply:
We will not allow extension of the ENUMS, but usage of a subset is appropriate. I will document accordingly.